A94-subchannel-otel-metrics.md

```diff
@@ -0,0 +1,128 @@
+
+If we end up discarding connection attempts as we do with the “happy eyeballs”
+algorithm (as per
+[A61: IPv4 and IPv6 Dualstack Backend Support](A61-IPv4-IPv6-dualstack-backends.md)),
+we should not record the connection attempt or the disconnection.
+
+Implementations that have already implemented the pick-first metrics should give
+enough time for users to transition to the new metrics. For example,
+implementations should report both the old pick-first metrics and the new
+subchannel metrics for 2 releases, and then remove the old pick-first metrics.
+
+Label Name              | Disposition | Description
+----------------------- | ----------- | -----------
+grpc.target             | Required    | Indicates the target of the gRPC channel (defined in [A66](A66-otel-stats.md).)
+grpc.lb.backend_service | Optional    | The backend service to which the RPC was routed (defined in [A89](https://github.com/grpc/proposal/pull/471))
+grpc.lb.locality        | Optional    | The locality to which the traffic is being sent. This will be set to the resolver attribute passed down from the weighted_target policy, or the empty string if the resolver attribute is unset. (defined in [A78](A78-grpc-metrics-wrr-pf-xds.md))
+grpc.disconnect_error   | Optional    | Reason for disconnection
+
+List of allowed values for `grpc.disconnect_error` -
+
+Error string         | Description
+-------------------- | -----------
+GOAWAY <ERROR_CODE>  | HTTP2 GOAWAY frame with error code for example (“GOAWAY NO_ERROR”, “GOAWAY PROTOCOL_ERROR”, “GOAWAY ENHANCE_YOUR_CALM”). The list of error codes is available in [RFC 9113](https://www.rfc-editor.org/rfc/rfc9113.html#name-error-codes).
+subchannel shutdown  | The subchannel was shutdown. This can happen due to reasons such as the parent channel shutting down, channel becoming idle, the load balancing policy changing due to a resolver update or a change in list of endpoint addresses.
+connection reset     | Connection was reset (eg. ECONNRESET, WSAECONNERESET)
+connection timed out | Connection timed out, also includes connections closed due to gRPC keepalives.
+connection aborted   | Connection was aborted
+socket error         | Any socket error not covered by “connection reset”, “connection timed out” and “connection aborted”. Implementations that are not able to differentiate between the different socket error codes should also use this.
+unknown              | Catch-all for all other reasons.
+
+For a given connection, there can be multiple reasons reported to the subchannel
+for disconnection. For example, a connection could have seen a GOAWAY frame with
+`ENHANCE_YOUR_CALM` and then a socket error Broken Pipe. In such cases, the
+first seen reason should be chosen, `GOAWAY ENHANCE_YOUR_CALM` in this case.
+
+We might add more error cases to this in the future.
+
+## Rationale
+
+### Renaming pick-first metrics
```

Comment:
Did we discuss what we're going to do with the existing PF metrics?  I guess we can't stop supporting them, but maybe we should mark them as deprecated somehow?

We should also probably add an "Updated by: gRFC 94" header to the top of A78.